File: src/com/Torvald/Terrarum/UserInterface/ConsoleWindow.java
package com.Torvald.Terrarum.UserInterface;

import com.Torvald.Terrarum.LangPack.Lang;
import com.Torvald.Terrarum.Terrarum;
import com.Torvald.Terrarum.ConsoleCommand.CommandInterpreter;
import com.Torvald.Terrarum.GameControl.Key;
import org.newdawn.slick.Color;
import org.newdawn.slick.GameContainer;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.Input;

/**
 * Created by minjaesong on 15-12-31.
 */
public class ConsoleWindow implements UICanvas {

    Color UIColour = new Color(0x90000000);

    private StringBuffer commandInputPool;
    private String prevCommand;

    private int inputCursorPos;

    private final int MESSAGES_MAX = 5000;
    private String[] messages;
    private int messageDisplayPos;
    private int messagesCount;

    private final int LINE_HEIGHT = 20;
    private final int MESSAGES_DISPLAY_COUNT = 9;

    int width;
    int height;

    public ConsoleWindow() {
        super();
        reset();
    }

    @Override
    public void update(GameContainer gc, int delta_t) {

    }

    @Override
    public void render(GameContainer gc, Graphics g) {
        g.setColor(UIColour);
        g.fillRect(0, 0, width, height);
        g.fillRect(0, 0, width, LINE_HEIGHT);

        String input = commandInputPool.toString();
        int inputDrawWidth = g.getFont().getWidth(input);
        int inputDrawHeight = g.getFont().getLineHeight();

        g.setColor(Color.white);
        g.drawString(input, 1, 0);
        g.fillRect(inputDrawWidth, 0, 2, inputDrawHeight);

        for (int i = 0; i < MESSAGES_DISPLAY_COUNT; i++) {
            String message = messages[messageDisplayPos + i];
            if (message != null) {
                g.drawString(message, 1, LINE_HEIGHT * (i + 1));
            }
        }
    }


    
    public void keyPressed(int key, char c) {
        // execute
        if (key == Key.RET && commandInputPool.length() > 0) {
            prevCommand = commandInputPool.toString();
            executeCommand();
            commandInputPool = new StringBuffer();
        }
        // backspace
        else if (key == Key.BKSP && commandInputPool.length() > 0) {
            commandInputPool.deleteCharAt(commandInputPool.length() - 1);
        }
        // get input
        else if ((key >= 2 && key <= 13)
                || (key >= 16 && key <= 27)
                || (key >= 30 && key <= 40)
                || (key >= 44 && key <= 53)
                || (commandInputPool.length() > 0 && key == 57)){
            commandInputPool.append(c);
            inputCursorPos += 1;
        }
        // prev command
        else if (key == Key.UP) {
            commandInputPool = new StringBuffer();
            commandInputPool.append(prevCommand);
        }
        // scroll up
        else if (key == Key.PGUP) {
            setDisplayPos(-MESSAGES_DISPLAY_COUNT + 1);
        }
        // scroll down
        else if (key == Key.PGDN) {
            setDisplayPos(MESSAGES_DISPLAY_COUNT - 1);
        }
    }

    @Override
    public void keyReleased(int key, char c) {

    }

    @Override
    public void mouseMoved(int oldx, int oldy, int newx, int newy) {

    }

    @Override
    public void mouseDragged(int oldx, int oldy, int newx, int newy) {

    }

    @Override
    public void mousePressed(int button, int x, int y) {

    }

    @Override
    public void mouseReleased(int button, int x, int y) {

    }

    @Override
    public void mouseWheelMoved(int change) {

    }

    @Override
    public void controllerButtonPressed(int controller, int button) {

    }

    @Override
    public void controllerButtonReleased(int controller, int button) {

    }

    @Override
    public void processInput(Input input) {

    }

    private void executeCommand() {
        sendMessage("> " + commandInputPool.toString());
        CommandInterpreter.execute(commandInputPool.toString());
    }

    public void sendMessage(String msg) {
        messages[messagesCount] = new String(msg);
        messagesCount += 1;
        if (messagesCount > MESSAGES_DISPLAY_COUNT) {
            messageDisplayPos = messagesCount - MESSAGES_DISPLAY_COUNT;
        }
    }

    private void setDisplayPos(int change) {
        int lowLim = 0;
        int highLim = (messagesCount <= MESSAGES_DISPLAY_COUNT)
                ? 0
                : messagesCount - MESSAGES_DISPLAY_COUNT;
        int newVal = messageDisplayPos + change;

        if (newVal < lowLim) {
            messageDisplayPos = lowLim;
        }
        else if (newVal > highLim) {
            messageDisplayPos = highLim;
        }
        else {
            messageDisplayPos = newVal;
        }

    }

    public void reset() {
        width = Terrarum.WIDTH;
        height = 200;

        messages = new String[MESSAGES_MAX];
        messageDisplayPos = 0;
        messagesCount = 0;
        inputCursorPos = 0;
        prevCommand = "";
        commandInputPool = new StringBuffer();

        if (Terrarum.game.auth.b()) sendMessage(Lang.get("DEV_MESSAGE_CONSOLE_CODEX"));
    }

    @Override
    public int getWidth() {
        return width;
    }

    @Override
    public int getHeight() {
        return height;
    }
}

File: src/net/torvald/terrarum/ui/ConsoleWindow.kt
package net.torvald.terrarum.ui

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.Camera
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import net.torvald.terrarum.App
import net.torvald.terrarum.App.printdbg
import net.torvald.terrarum.Terrarum
import net.torvald.terrarum.TerrarumAppConfiguration
import net.torvald.terrarum.ccE
import net.torvald.terrarum.console.Authenticator
import net.torvald.terrarum.console.CommandInterpreter
import net.torvald.terrarum.gameactors.AVKey
import net.torvald.terrarum.gamecontroller.TerrarumKeyboardEvent
import net.torvald.terrarum.langpack.Lang
import net.torvald.unicode.EMDASH
import net.torvald.util.CircularArray


/**
 * Don't let the debug console have the ```toggleKeyLiteral```, this must open even when the game is paused
 *
 * Created by minjaesong on 2015-12-31.
 */
class ConsoleWindow : UICanvas() {

    internal var UIColour = Color(0x404080_80.toInt())

    private var inputCursorPos: Int = 0
    private val MESSAGES_MAX = 1000
    private val COMMAND_HISTORY_MAX = 100
    private var messages = CircularArray<String>(MESSAGES_MAX, true)
    private var messageDisplayPos: Int = 0
    private var messagesCount: Int = 0

//    private var commandInputPool: StringBuilder? = null
    private var commandHistory = CircularArray<String>(COMMAND_HISTORY_MAX, true)

    private val LINE_HEIGHT = 20
    private val MESSAGES_DISPLAY_COUNT = 11

    private val inputToMsgboxGap = 3

    override var width: Int = App.scr.width
    override var height: Int = LINE_HEIGHT * (MESSAGES_DISPLAY_COUNT + 1) + inputToMsgboxGap

    override var openCloseTime = 0f

    private var drawOffX: Float = 0f
    private var drawOffY: Float = -height.toFloat()
    private var openingTimeCounter = 0f

    private var historyIndex = -1

    private var iMadeTheGameToPause = false

    private val textinput = UIItemTextLineInput(this, 0, 0, this.width, keyFilter = { e ->
        !e.keycodes.contains(Input.Keys.GRAVE)
    })

    private var clickLatched = false

    init {
        reset()
        addUIitem(textinput)
        textinput.isEnabled = false
    }

    private val lb = ArrayList<String>()

    override fun updateUI(delta: Float) {
        Terrarum.ingame?.let {
            if (Authenticator.b()) {
                lb.clear()

                val actorsUnderCursor = it.getActorsAt(Terrarum.mouseX, Terrarum.mouseY)
                actorsUnderCursor.filter { it.referenceID < 2147483647 }.forEach { // filter out the BlockMarkerActor
                    lb.add("${it.referenceID} (${it.actorValue[AVKey.NAME] ?: "\u03AF-${it.javaClass.simpleName}"})")
                }

                it.setTooltipMessage(if (lb.size > 0) lb.joinToString("\n") else null)

                // click to enter the actor's reference ID
                if (lb.size > 0 && !clickLatched && Gdx.input.isButtonPressed(App.getConfigInt("config_mouseprimary"))) {
                    clickLatched = true
                    textinput.appendText(lb.first().substringBefore(' '))
                }
            }
            else {
                it.setTooltipMessage(null)
            }
        }

        uiItems.forEach { it.update(delta) }


        if (!Gdx.input.isButtonPressed(App.getConfigInt("config_mouseprimary"))) {
            clickLatched = false
        }

        textinput.isEnabled = (isOpened && !isClosing)
    }

    override fun renderUI(batch: SpriteBatch, camera: OrthographicCamera) {
        // background
        batch.color = UIColour
        Toolkit.fillArea(batch, drawOffX, drawOffY, width.toFloat(), height.toFloat())
        Toolkit.fillArea(batch, drawOffX, drawOffY, width.toFloat(), LINE_HEIGHT.toFloat())

//        val input = commandInputPool!!.toString()
//        val inputDrawWidth = App.fontGame.getWidth(input)
//        val inputDrawHeight = App.fontGame.lineHeight
//
//         text and cursor
//        batch.color = Color.WHITE
//        App.fontGame.draw(batch, input, 1f + drawOffX, drawOffY)
//
//        batch.color = Color(0x7f7f7f_ff)
//        Toolkit.fillArea(batch, inputDrawWidth.toFloat() + drawOffX + 1, drawOffY, 2f, inputDrawHeight)
//        batch.color = Color.WHITE
//        Toolkit.fillArea(batch, inputDrawWidth.toFloat() + drawOffX + 1, drawOffY, 1f, inputDrawHeight - 1)


        // messages
        batch.color = Color.WHITE

        for (i in 0 until MESSAGES_DISPLAY_COUNT) {
            val message = messages[messageDisplayPos + i] ?: ""
            App.fontGame.draw(batch, message, 1f + drawOffX, (LINE_HEIGHT * (MESSAGES_DISPLAY_COUNT - i)).toFloat() + drawOffY + inputToMsgboxGap)
        }

        uiItems.forEach { it.render(batch, camera) }
    }

    override fun inputStrobed(e: TerrarumKeyboardEvent) {
        uiItems.forEach { it.inputStrobed(e) }
    }

    override fun keyDown(key: Int): Boolean {
        try {
            val textOnBuffer = textinput.getText()

            // history
            if (key == Input.Keys.UP && historyIndex < commandHistory.elemCount)
                historyIndex++
            else if (key == Input.Keys.DOWN && historyIndex >= 0)
                historyIndex--
            else if (key != Input.Keys.UP && key != Input.Keys.DOWN)
                historyIndex = -1

            // execute
            if (key == Input.Keys.ENTER && textOnBuffer.isNotEmpty()) {
                commandHistory.appendHead(textOnBuffer)
                executeCommand(textOnBuffer)
                textinput.clearText()
            }
            // scroll
            else if (key == Input.Keys.UP || key == Input.Keys.DOWN) {
                // create new stringbuilder
                textinput.clearText()
                if (historyIndex >= 0) // just leave blank if index is -1
                    textinput.setText(commandHistory[historyIndex] ?: "")
            }
            // delete input
//        else if (key == Input.Keys.BACKSPACE) {
//            commandInputPool = StringBuilder()
//        }
            // message scroll up
            else if (key == Input.Keys.PAGE_UP) {
                setDisplayPos(-MESSAGES_DISPLAY_COUNT + 1)
            }
            // message scroll down
            else if (key == Input.Keys.PAGE_DOWN) {
                setDisplayPos(MESSAGES_DISPLAY_COUNT - 1)
            }
        }
        catch (e: ConcurrentModificationException) {
            System.err.println("[ConsoleWindow.keyDown] ConcurrentModificationException with key ${Input.Keys.toString(key)}")
        }


        return true
    }

    val acceptedChars = "1234567890-=qwfpgjluy;[]\\arstdhneio'zxcvbkm,./!@#$%^&*()_+QWFPGJLUY:{}|ARSTDHNEIO\"ZXCVBKM<>? ".toSet()

    /*override fun keyTyped(character: Char): Boolean {
        if (character in acceptedChars) {
            commandInputPool!!.append(character)
            inputCursorPos += 1

            return true
        }
        else {
            return false
        }
    }*/

    override fun keyUp(keycode: Int): Boolean {
        return false
    }

    private fun executeCommand(s: String) {
        CommandInterpreter.execute(s)
    }

    fun sendMessage(msg: String) {
        messages.appendHead(msg)
        messagesCount += 1

    }

    private fun setDisplayPos(change: Int) {
        val lowLim = 0
        val highLim = if (messagesCount <= MESSAGES_DISPLAY_COUNT)
            0
        else
            messagesCount - MESSAGES_DISPLAY_COUNT
        val newVal = messageDisplayPos + change

        if (newVal < lowLim) {
            messageDisplayPos = lowLim
        }
        else if (newVal > highLim) {
            messageDisplayPos = highLim
        }
        else {
            messageDisplayPos = newVal
        }

    }

    fun reset() {
        messages = CircularArray<String>(MESSAGES_MAX, true)
        messageDisplayPos = 0
        messagesCount = 0
        inputCursorPos = 0
        commandHistory = CircularArray<String>(COMMAND_HISTORY_MAX, true)
        textinput.clearText()

        if (Authenticator.b()) {
            sendMessage("$ccE${TerrarumAppConfiguration.GAME_NAME} ${App.getVERSION_STRING()} $EMDASH ${TerrarumAppConfiguration.COPYRIGHT_DATE_NAME}")
            sendMessage("$ccE${TerrarumAppConfiguration.COPYRIGHT_LICENSE}")
            sendMessage(Lang["DEV_MESSAGE_CONSOLE_CODEX"])
        }
    }

    override fun doOpening(delta: Float) {
        Terrarum.ingame?.let {
//            printdbg(this, "Game was paused beforehand: ${it.paused}")
            if (!it.paused) {
                iMadeTheGameToPause = true
                it.pause()
            }
            else {
                iMadeTheGameToPause = false
            }
//            printdbg(this, "I made the game to pause: $iMadeTheGameToPause")
        }
        /*openingTimeCounter += delta
        drawOffY = MovementInterpolator.fastPullOut(openingTimeCounter.toFloat() / openCloseTime.toFloat(),
                -height.toFloat(), 0f
        )*/
    }

    override fun doClosing(delta: Float) {
        /*openingTimeCounter += delta
        drawOffY = MovementInterpolator.fastPullOut(openingTimeCounter.toFloat() / openCloseTime.toFloat(),
                0f, -height.toFloat()
        )*/
        textinput.isEnabled = false
        textinput.mouseoverUpdateLatch = false
    }

    override fun endOpening(delta: Float) {
        drawOffY = 0f
        openingTimeCounter = 0f
        textinput.isEnabled = true
        textinput.mouseoverUpdateLatch = true
    }

    override fun endClosing(delta: Float) {
//        printdbg(this, "Close -- I made the game to pause: $iMadeTheGameToPause")
        if (iMadeTheGameToPause) {
            Terrarum.ingame?.resume()
//            printdbg(this, "Close -- resume game")
        }
        iMadeTheGameToPause = false
        Terrarum.ingame?.setTooltipMessage(null)
        drawOffY = -height.toFloat()
        openingTimeCounter = 0f
    }

    override fun dispose() {
        uiItems.forEach { it.dispose() }
    }
}
